### DISCUSSION TITLE
Problems with using the generated FMU - fails to instansiate FMU (fmi3InstantiateCoSimulation)

### DISCUSSION THREAD

**Discussion by mads1247** [StephenSmith25/PythonFMU3#4] [Q&A]
Hello,

I have generated a FMU and i'm not able to run it with Matlab or FMPy GUI. When I run a custom script using FMPy, i get the expected results. When trying to run the FMU with Matlab of FMPy GUI it says that there is some error with the fmi3InstantiateCoSimulation. But what I don't understand is why it then works with the custom script where I run the .instantiate() function on the FMU that (as far as I can se in the documentation) also run the fmi3InstantiateCoSimulation works, but dosen't work elsewhere. 

FMPy GUI Bug log:
`fmi3InstantiateCoSimulation(instanceName="PythonSlave", instantiationToken="834f38c2-d15c-11ee-abb3-98597af38661", resourcePath="C:\Users\mela\AppData\Local\Temp\tmpakwicwne\resources\", visible=True, loggingOn=True, eventModeUsed=False, earlyReturnAllowed=False, requiredIntermediateVariables=NULL, nRequiredIntermediateVariables=0, instanceEnvironment=0x0, logMessage=<CFunctionType object at 0x0000029857A8DCC0>, intermediateUpdate=<CFunctionType object at 0x000002987F9EB520>) -> 0x0` :

Custom script I test the code with: 
`from fmpy import read_model_description, extract
from fmpy.fmi3 import FMU3Slave
from fmpy.util import plot_result, download_test_file
import numpy as np
import shutil
import math

def simulate_custom_input(show_plot=True):

    # define the model name and simulation parameters
    fmu_filename = 'PythonSlave.fmu'
    start_time = 0.0
    threshold = 2.0
    stop_time = 2.0
    step_size = 0.05

    # read the model description
    model_description = read_model_description(fmu_filename)

    print(model_description)

    # collect the value references
    vrs = {}
    for variable in model_description.modelVariables:
        vrs[variable.name] = variable.valueReference

    print(vrs)

    # get the value references for the variables we want to get/set
    vr_input   = vrs['q']      # normalized force on the 3rd clutch
    vr_output = vrs['tau']  # angular velocity of the 4th inertia

    # extract the FMU
    unzipdir = extract(fmu_filename)

    fmu = FMU3Slave(guid=model_description.guid,
                    unzipDirectory=unzipdir,
                    modelIdentifier=model_description.coSimulation.modelIdentifier,
                    instanceName='instance1')

    # initialize
    fmu.instantiate()
    fmu.enterInitializationMode()
    fmu.exitInitializationMode()

    time = start_time

    rows = []  # list to record the results

    # simulation loop
    while time < stop_time:

        # NOTE: the FMU.get*() and FMU.set*() functions take lists of
        # value references as arguments and return lists of values

        # set the input
        fmu.setFloat64([vr_input], [0.0 if time < 0.9 else math.pi/2])

        # perform one step
        fmu.doStep(currentCommunicationPoint=time, communicationStepSize=step_size)

        # advance the time
        time += step_size

        # get the values for 'inputs' and 'outputs[4]'
        inputs, output = fmu.getFloat64([vr_input, vr_output])

        # append the results
        rows.append((time, inputs, output))

    fmu.terminate()
    fmu.freeInstance()

    # clean up
    shutil.rmtree(unzipdir, ignore_errors=True)

    # convert the results to a structured NumPy array
    result = np.array(rows, dtype=np.dtype([('time', np.float64), ('q', np.float64), ('tau', np.float64)]))
    print(result)

if __name__ == '__main__':

    simulate_custom_input()`

**Comment by StephenSmith25** [comment]
Hi,

It’s strange that the FMU works with your script, but not with the fmpy GUI. Just a couple of questions

1)      Are you using the same python version to run both the script and the fmpy gui?

2)      What platform are you on?

3)      Which python version are you using?

4)      How did you install pythonFMU3?

I wonder if the discussions in https://github.com/NTNU-IHB/PythonFMU/issues/110 are of any use?

Have you tried using just the simple example in https://github.com/StephenSmith25/PythonFMU3/blob/master/examples/demoslave.py. You should be able to run this with both the GUI, using

```
python -m fmpy.gui
```

And a simple script,

```

from fmpy import *

from fmpy.util import plot_result

fmu = 'NAME OF FMU’

input_values  = {}

dump(fmu)

result = simulate_fmu(fmu, start_time=0, stop_time=1.5, step_size=1e-3, fmi_call_logger=lambda s: print('[FMI] ' + s), start_values=input_values)

plot_result(result)

```

Can you share more about the FMU you are running and whether it has any dependencies?

**Comment by mads1247** [comment]
Okay so the problem with FMPy was, as you suggested, that I was running in 2 different environments. But I am still not able to run it with Simulink. 

I have tried to generate the example above (the Demoslave) and here it works. But when I try to import numpy, it either crashes Matlab, or it fails to instasiate the FMU. I have tried using the PythonFMU to generate with the 2.0 standard and it has the same issue. 
Is this something you have heard of before?

**Reply by StephenSmith25** [reply]
Unfortunately, loading the FMU within Matlab is difficult. If it works without numpy then I would guess the issue is with the python library it finds. You can try two things if you are on Linux

1. You can modify LD_LIBRARY_PATH to add folder which contains libpython (the one with numpy). Then launch Matlab

2. You can use LD_PRELOAD before you launch Matlab to load the python library first. E.g LD_PRELOAD=/PATH/TO/PYTHON/libpython.so

If you need more info let me know. Unfortunately, I am not a Matlab user, so I cannot speak from personal experience. Also windows will likely require something different.
